statsmodels/stats/contingency_tables.py

```diff
@@ -0,0 +1,1215 @@
+
+
+def _handle_pandas_square(table):
+    """
+    Reindex a pandas table so that it becomes square (extending the
+    row and column index as needed).
+    """
+
+    if not isinstance(table, pd.DataFrame):
+        return table
+
+    # If the table is not square, make it square
+    if table.shape[0] != table.shape[1]:
+        ix = list(set(table.index) | set(table.columns))
+        table = table.reindex(ix, axis=0)
+        table = table.reindex(ix, axis=1)
+
+    # Ensures that the roes and columns are in the same order.
+    table = table.reindex(table.columns)
+
+    return table
+
+
+class _bunch(object):
+    pass
+
+
+class Table(object):
+    """
+    Analyses that can be performed on a two-way contingency table.
+
+    Parameters
+    ----------
+    table : array-like
+        A contingency table.
+    shift_zeros : boolean
+        If True and any cell count is zero, add 0.5 to all values
+        in the table.
+
+    See also
+    --------
+    statsmodels.graphics.mosaicplot.mosaic
+    scipy.stats.chi2_contingency
+    """
+
+    def __init__(self, table, shift_zeros=True):
+
+        self.table_orig = table
+        self.table = np.asarray(table, dtype=np.float64)
+
+        if shift_zeros and (self.table.min() == 0):
+            self.table = self.table + 0.5
+
+
+    @classmethod
+    def from_data(cls, data):
+        """
+        Construct a Table object from data.
+
+        Parameters
+        ----------
+        data : array-like
+            The raw data, from which a cross-table is constructed
+            using the first two columns.
+
+        Returns
+        -------
+        A Table instance.
+        """
+
+        if isinstance(data, pd.DataFrame):
+            table = pd.crosstab(data.iloc[:, 0], data.iloc[:, 1])
+        else:
+            table = pd.crosstab(data[:, 0], data[:, 1])
+
+        return cls(table)
+
+
+    @cache_readonly
+    def nominal_association(self):
+        """
+        Assess independence between rows and columns using chi^2 test.
+
+        Returns
+        -------
+        stat : float
+            The chi^2 test statistic.
+        df : integer
+            The degrees of freedom of the reference distribution
+        pvalue : float
+            The p-value for the test.
+        """
+        stat = np.asarray(self.chi2_contribs).sum()
+        df = np.prod(np.asarray(self.table.shape) - 1)
+        pvalue = 1 - stats.chi2.cdf(stat, df)
+        b = _bunch()
+        b.stat = stat
+        b.df = df
+        b.pvalue = pvalue
+        return b
+
+
+    def ordinal_association(self, row_scores=None, col_scores=None):
+        """
+        Assess independence between rows and columns for ordinal factors.
+
+        This is the 'linear by linear' association test.
+
+        Parameters
+        ----------
+        row_scores : array-like
+            An array of numeric row scores
+        col_scores : array-like
+            An array of numeric column scores
```

Comment:
What are "scores"?   I still have problems with different meanings of that word in different contexts.
Are these counts in the context of contingency tables?